lib/SILOptimizer/Transforms/StringOptimization.cpp

```diff
@@ -0,0 +1,475 @@
+    changed |= optimizeBlock(block);
+  }
+  return changed;
+}
+
+/// Run the optimization on a basic block.
+bool StringOptimization::optimizeBlock(SILBasicBlock &block) {
+  bool changed = false;
+  
+  /// Maps identifyable objects (alloc_stack, inout parameters) to string values
+  /// which are stored in those objects.
+  llvm::DenseMap<SILValue, SILValue> storedStrings;
+  
+  for (auto iter = block.begin(); iter != block.end();) {
+    SILInstruction *inst = &*iter++;
+
+    if (StoreInst *store = isStringStoreToIdentifyableObject(inst)) {
+      storedStrings[store->getDest()] = store->getSrc();
+      continue;
+    }
+    if (ApplyInst *append = isSemanticCall(inst, semantics::STRING_APPEND, 2)) {
+      if (optimizeStringAppend(append, storedStrings)) {
+        changed = true;
+        continue;
+      }
+    }
+    if (ApplyInst *typeName = isSemanticCall(inst, semantics::TYPENAME, 2)) {
+      if (optimizeTypeName(typeName)) {
+        changed = true;
+        continue;
+      }
+    }
+    // Remove items from storedStrings if inst overwrites (or potentially
+    // overwrites) a stored String in an identifyable object.
+    invalidateModifiedObjects(inst, storedStrings);
+  }
+  return changed;
+}
+
+/// Optimize String.append in case anything is known about the parameters.
+bool StringOptimization::optimizeStringAppend(ApplyInst *appendCall,
+                            llvm::DenseMap<SILValue, SILValue> &storedStrings) {
+  SILValue rhs = appendCall->getArgument(0);
+  StringInfo rhsString = getStringInfo(rhs);
+  
+  // Remove lhs.append(rhs) if rhs is empty.
+  if (rhsString.isEmpty()) {
+    appendCall->eraseFromParent();
+    return true;
+  }
+  
+  SILValue lhsAddr = appendCall->getArgument(1);
+  StringInfo lhsString = getStringInfo(storedStrings[lhsAddr]);
+
+  // The following two optimizations are a trade-off: Performance-wise it may be
+  // benefitial to initialize an empty string with reserved capacity and then
+  // append multiple other string components.
+  // Removing the empty string (with the reserved capacity) might result in more
+  // allocations.
+  // So we just do this optimization up to a certain capacity limit (found by
+  // experiment).
+  if (lhsString.reservedCapacity > 50)
+    return false;
+
+  // Replace lhs.append(rhs) with 'lhs = rhs' if lhs is empty.
+  if (lhsString.isEmpty()) {
+    replaceAppendWith(appendCall, rhs, /*copyNewValue*/ true);
+    storedStrings[lhsAddr] = rhs;
+    return true;
+  }
+  
+  // Replace lhs.append(rhs) with "lhs = lhs + rhs" if both lhs and rhs are
+  // constant.
+  if (lhsString.isConstant() && rhsString.isConstant()) {
+    std::string concat = lhsString.str;
+    concat += rhsString.str;
+    if (ApplyInst *stringInit = createStringInit(concat, appendCall)) {
+      replaceAppendWith(appendCall, stringInit, /*copyNewValue*/ false);
+      storedStrings[lhsAddr] = stringInit;
+      return true;
+    }
+  }
+  
+  return false;
+}
```

Comment:
I was wondering if there are any dead operands after this optimization, whether they all will be cleaned up by a later dead code elimination.

Reply:
I have noticed that String.makeUTF8 init may not get DCE'ed some times. Otherwise, there is a `InstructionDeleter` utility that can deleted instructions and track their operands and can DCE them later.

Reply:
It should always be cleaned up because it's defined as "readonly". InstructionDeleter is not really required here, because this is done multiple times in following passes.